zeroknowledgeproof/src/main/java/com/zeroknowledgeproof/RangeProof.kt

```diff
@@ -0,0 +1,220 @@
+    val rand = SecureRandom(); // SEKURE random
+
+    fun Prover (m: Int, a: Int, b: Int) : Boolean {
+        // Prove that boundedNumber is between lower and upper bound
+        if ((m - a + 1)*(b - m + 1) <= 0) {
+            throw IllegalArgumentException("Cannot generate proof, because this is just NOT TRUE")
+        }
+
+
+        val w = rand.nextInt(10000)
+
+        val theProof = toBigInt(w * w).times(toBigInt(m - a + 1)).times(toBigInt(b - m + 1))
+
+        val m4 = generateToSum(toBigInt(sqrt(theProof.toDouble()).toInt()))
+        val m3 = m4.times(m4);
+        val m2 = generateToSum(theProof.subtract(m3)) //Take a random int of interval [0, theProof-m3]
+        val m1 = theProof.subtract(m3).subtract(m2)      // Now m1 + m2 + m3 = w^2 * (m-a+1) * (b-m+1)
+
+//        var P: BigInteger
+//        var Q: BigInteger
+        var p: BigInteger
+        var q: BigInteger
+//        var PminOne: BigInteger
+
+//        var attempts = 0;
+//        do {
+//            P = BigInteger.probablePrime(2048, rand)
+//            attempts++
+//            if (attempts % 25 == 0) {
+//                println("Calculating P and p, now at " + attempts + " attempts.")
+//            }
+//            PminOne = P.subtract(ONE).divide(BigInteger.valueOf(2))
+//        } while (!PminOne.isProbablePrime(50) )
+
+          var QPprimes = generateSafePrimes(2048, 50)
+            val P = QPprimes[0]
+            val Q = QPprimes[1]
+
+//        do {
+//            Q = BigInteger.probablePrime(2048, rand)
+//            attempts++
+//            if (attempts % 25 == 0) {
+//                println("Calculating Q and q now at " + attempts + " attempts.")
+//            }
+//        } while (!Q.subtract(ONE).divide(BigInteger.valueOf(2)).isProbablePrime(50) || P.equals(Q))
+
+        val N = P.multiply(Q)
+
+        val k1 = BigInteger(1024, rand)
+        val k2 = BigInteger(160, rand)
+
+
+        val (g,h) = findGenerators(findTwoPrimes(2048))
+
+        val r = BigInteger(N.bitLength(), rand)
+
+        val c = g.modPow(toBigInt(m), N).times(h.modPow(r, N)).mod(N)
+
+        val c1 = c.divide(g.modPow(toBigInt(a-1), N)).mod(N) // c1 = c/(g^a-1) modN
+        //TODO Maybe this should not be Integer division, but cyclic group division (groot crypto fun)
+        val c2 = g.modPow(toBigInt(b+1), N).divide(c).mod(N)
+        val rPrime = BigInteger(N.bitLength(), rand)
+
+        val cPrime = c1.modPow(toBigInt(b - m + 1), N).times(h.modPow(rPrime, N)).mod(N)
+        val rDPrime = BigInteger(N.bitLength(), rand)
+
+        val cDPrime = cPrime.modPow(toBigInt(w*w), N).times(h.modPow(rDPrime, N)).mod(N)
+
+        val publicProof = cPrime.toString(10) + "," + h.toString(10)
+        print("The Public Proof is " + publicProof)
+
+        // Prove that the hidden number is a square with base (c', h)
+        //TODO: implement, ING used a lot of classes for this
+
+        val rSum = toBigInt((w*w)*((b-m+1))).times(r) + rPrime + rDPrime
```

Comment:
beware rsum should be : w^2 * ((b - m + 1)r  + r') + rDprime

Reply:
This is part of rangeproof and should not even be in this pr.